File: sdks/python/src/oap_sdk/subscription.py
"""Manifest Subscriptions (RFC 0022)."""

from __future__ import annotations

from dataclasses import dataclass
from typing import List, Optional

from .signing import canonicalize, sha256_hex


@dataclass
class ManifestSubscription:
    id: str
    subscriber_did: str
    target_tool_did: str
    callback_url: str
    watched_fields: List[str]
    created_at: str


@dataclass
class ManifestUpdateNotification:
    subscription_id: str
    target_tool_did: str
    manifest_version: str
    changed_fields: List[str]
    prior_manifest_hash: str
    new_manifest_hash: str
    issued_at: str
    signature: Optional[str] = None


def verify_manifest_update(notification: ManifestUpdateNotification, new_manifest: dict) -> bool:
    """Return True if the canonicalized manifest hashes to the value the notification claims."""
    return sha256_hex(canonicalize(new_manifest)) == notification.new_manifest_hash

File: sdks/typescript/src/subscription.ts
/**
 * Manifest Subscriptions (RFC 0022).
 *
 * Helper to subscribe to a remote tool's manifest changes via the
 * implementation's `subscribe` endpoint, and to validate incoming
 * manifest-update notifications.
 */

import { canonicalize, sha256Hex } from './signing.js';

export interface ManifestSubscription {
  subscriber_did: string;
  target_tool_did: string;
  callback_url: string;
  watched_fields?: string[];
  created_at: string;
  id: string;
}

export interface ManifestUpdateNotification {
  subscription_id: string;
  target_tool_did: string;
  manifest_version: string;
  changed_fields: string[];
  prior_manifest_hash: string;
  new_manifest_hash: string;
  issued_at: string;
  signature?: string;
}

export async function subscribeToManifest(opts: {
  targetToolUrl: string;
  subscriberDid: string;
  callbackUrl: string;
  watchedFields?: string[];
}): Promise<ManifestSubscription> {
  const url = `${opts.targetToolUrl.replace(/\/$/, '')}/oap/subscribe`;
  const res = await fetch(url, {
    method: 'POST',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify({
      subscriber_did: opts.subscriberDid,
      callback_url: opts.callbackUrl,
      watched_fields: opts.watchedFields,
    }),
  });
  if (!res.ok) throw new Error(`subscribe failed: ${res.status}`);
  return (await res.json()) as ManifestSubscription;
}

export function verifyManifestUpdate(
  notification: ManifestUpdateNotification,
  newManifest: unknown,
): boolean {
  const computed = sha256Hex(canonicalize(newManifest));
  return computed === notification.new_manifest_hash;
}
